test/src/device-state/current-state.spec.ts

```diff
@@ -0,0 +1,585 @@
+				'update_pending',
+				'update_downloaded',
+				'update_failed',
+				'status',
+			]);
+			expect(getStateDiff()).to.deep.equal({
+				local: _.pick(testDeviceReportFields, ['os_version']),
+				dependent: _.pick(testDeviceReportFields, ['status']),
+			});
+		});
+	});
+
+	describe('throttled report', () => {
+		const report = currentState.__get__('report');
+
+		let configGetManyStub: SinonStub;
+
+		before(() => {
+			configGetManyStub = stub(config, 'getMany');
+			requestInstance.patchAsync = () =>
+				Bluebird.resolve([{ statusCode: 200 }]);
+		});
+
+		after(() => {
+			configGetManyStub.restore();
+		});
+
+		beforeEach(() => {
+			configGetManyStub.resolves(
+				_.omit(testDeviceConf, ['deviceId', 'disableHardwareMetrics']) as any,
+			);
+		});
+
+		afterEach(() => {
+			// Clear the throttle time limit between tests
+			report.cancel();
+		});
+
+		it("doesn't report if current state hasn't changed", async () => {
+			// A beforeEach hook resets the global state objects to all be empty, so
+			// by default, report() will return 0 in this test env
+			expect(await report()).to.equal(0);
+		});
+
+		it('errors when provided invalid uuid or apiEndpoint', async () => {
+			configGetManyStub.resolves(
+				_.omit(testDeviceConf, [
+					'deviceId',
+					'disableHardwareMetrics',
+					'uuid',
+					'apiEndpoint',
+				]) as any,
+			);
+			stateForReport.local = { ...testDeviceReportFields };
+			await expect(report()).to.be.rejectedWith(InternalInconsistencyError);
+		});
+
+		it('resets stateReportErrors to 0 on patch success', async () => {
+			spy(_, 'assign');
+
+			stateForReport.local = { ...testDeviceReportFields };
+			await report();
+
+			expect(stateReportErrors).to.equal(0);
+			expect(_.assign as SinonSpy).to.have.been.calledTwice;
+
+			(_.assign as SinonSpy).restore();
+		});
+
+		it('handles errors on state patch failure', async () => {
+			// Overwrite default patchAsync response to return erroring statusCode
+			patchAsyncSpy.restore();
+			requestInstance.patchAsync = () =>
+				Bluebird.resolve([{ statusCode: 400 }]);
+			patchAsyncSpy = spy(requestInstance, 'patchAsync');
+
+			stub(log, 'error');
+
+			stateForReport.local = { ...testDeviceReportFields };
+			await report();
+
+			expect((log.error as SinonStub).lastCall.args[0]).to.equal(
+				'Non-200 response from the API! Status code: 400 - message:',
+			);
+			(log.error as SinonStub).restore();
+
+			// Reset to default patchAsync
+			requestInstance.patchAsync = () =>
+				Bluebird.resolve([{ statusCode: 200 }]);
+		});
+	});
+
+	describe('reportCurrentState', () => {
+		const reportCurrentState = currentState.__get__('reportCurrentState');
+		const report = currentState.__get__('report');
+		let localDisableHardwareMetrics: any;
+		const localAppUpdatePollInterval = 1000;
+
+		const unhandledRejectionHandler = () => {
+			/* noop */
+		};
+
+		before(() => {
+			stub(deviceState, 'getStatus').resolves({});
```

Comment:
We should make a note to improve this module. The fact that so many stubs are necessary and stuff like the sleep and wait for the function to be called tell us that there is something wrong with the architecture

Reply:
Agreed. The stubs for fine and expected for unit testing, but having to `sleep(200)` for a function call to complete because it's the only way to test its effects isn't great.